### PR TITLE
feat: generate test wraps for html components

### LINKED ISSUE #1470: Add test wrappers for HTML components

### PR DESCRIPTION
**Pull request by caalador**
Add test wrap for all html components

fixes #1470

### CODE REVIEW

**Comment by caalador** [comment]
Missing test cases for all html components.

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/AnchorTester.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+     * Wrap given component for testing.
+     *
+     * @param component
+     *            target component
+     */
+    public AnchorTester(Anchor component) {
+        super(component);
+    }
+
+    public String getHref() {
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/AnchorTester.java:33]: javadoc from Anchor component for public method

```suggestion
    /**
     * Gets the URL that the anchor links to.
     *
     * @return the href value, or <code>""</code> if no href has been set
     */
    public String getHref() {
```

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/ClickHandler.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ * This program is available under Commercial Vaadin Developer License
+ * 4.0 (CVDLv4).
+ *
+ *
+ * For the full License, see <https://vaadin.com/license/cvdl-4.0>.
+ */
+package com.vaadin.flow.component.html.testbench;
+
+import com.vaadin.flow.component.ClickEvent;
+import com.vaadin.flow.component.ComponentUtil;
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/ClickHandler.java:13]: Unused imports
```suggestion
```

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/AnchorTester.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ */
+package com.vaadin.flow.component.html.testbench;
+
+import java.lang.reflect.Field;
+
+import com.vaadin.flow.component.HasElement;
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.component.html.Anchor;
+import com.vaadin.flow.router.RouteConfiguration;
+import com.vaadin.flow.server.AbstractStreamResource;
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/AnchorTester.java:18]: Unused import
```suggestion
```

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/HtmlComponentTester.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+     * @throws IllegalStateException
+     *             if not title has been set
+     */
+    public String getTitle() {
+        ensureVisible();
+        return getComponent().getTitle()
+                .orElseThrow(() -> new IllegalStateException(
+                        "No title set for " + getComponent().getClassName()));
+    }
+
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/HtmlComponentTester.java:40]: Should we maybe add a `getText()` also for this Tester, delegating to `getElement().getTextRecursively()`?

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/DescriptionListTester.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+     * @return
+     */
+    public List<DescriptionList.Description> getDesciptions() {
+        return getComponent().getChildren()
+                .filter(DescriptionList.Description.class::isInstance)
+                .map(DescriptionList.Description.class::cast)
+                .collect(Collectors.toList());
+    }
+
+    public List<DescriptionList.Term> getTerms() {
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/DescriptionListTester.java:42]:
```suggestion
    /**
     * Get all terms of DescriptionList.
     *
     * @return the list of terms for this description list
     */
     public List<DescriptionList.Term> getTerms() {
```

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/DescriptionListTester.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+     *            target component
+     */
+    public DescriptionListTester(DescriptionList component) {
+        super(component);
+    }
+
+    /**
+     * Get all descriptions of DescriptionList.
+     *
+     * @return
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/DescriptionListTester.java:33]:
```suggestion
     * @return the list of descriptions for this description list
```

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/InputTester.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+ */
+package com.vaadin.flow.component.html.testbench;
+
+import org.slf4j.LoggerFactory;
+
+import com.vaadin.flow.component.html.Input;
+import com.vaadin.flow.component.shared.HasClearButton;
+import com.vaadin.flow.component.textfield.TextFieldTester;
+import com.vaadin.testbench.unit.ComponentTester;
+import com.vaadin.testbench.unit.Tests;
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/InputTester.java:18]: Unused imports
```suggestion
import com.vaadin.flow.component.html.Input;
import com.vaadin.testbench.unit.ComponentTester;
import com.vaadin.testbench.unit.Tests;
```

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        super(component);
+    }
+
+    /**
+     * Get all ListItems for unordered list.
+     *
+     * @return list of items
+     */
+    public List<ListItem> getItems() {
+        return getComponent().getChildren().filter(ListItem.class::isInstance)
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java:37]: Should we check that the component is visible?

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                .map(ListItem.class::cast).collect(Collectors.toList());
+    }
+
+    /**
+     * Get all ListItems for unordered list already wrapped for testing.
+     *
+     * @return list of pre-wrapped items
+     */
+    public List<ListItemTester> getWrappedItems() {
+        return getComponent().getChildren().filter(ListItem.class::isInstance)
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java:47]: Should we check that the component is visible?

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/UnorderedListTester.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        super(component);
+    }
+
+    /**
+     * Get all ListItems for unordered list.
+     *
+     * @return list of items
+     */
+    public List<ListItem> getItems() {
+        return getComponent().getChildren().filter(ListItem.class::isInstance)
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/UnorderedListTester.java:37]: Should we check that the component is visible?

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/UnorderedListTester.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+                .map(ListItem.class::cast).collect(Collectors.toList());
+    }
+
+    /**
+     * Get all ListItems for unordered list already wrapped for testing.
+     *
+     * @return list of pre-wrapped items
+     */
+    public List<ListItemTester> getWrappedItems() {
+        return getComponent().getChildren().filter(ListItem.class::isInstance)
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/UnorderedListTester.java:47]: Should we check that the component is visible?

File: vaadin-testbench-unit/src/test/java/com/vaadin/flow/component/html/testbench/AnchorTesterTest.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        UI.getCurrent().add(anchor);
+
+        final IllegalStateException exception = assertThrows(
+                IllegalStateException.class, () -> test(anchor).click());
+
+        Assertions.assertEquals("Anchor target seems to be a resource",
+                exception.getMessage());
+    }
+
+}
```
Comment (mcollovati) [inline vaadin-testbench-unit/src/test/java/com/vaadin/flow/component/html/testbench/AnchorTesterTest.java:73]: newline missing
```suggestion
}

```

**Review by mcollovati — CHANGES_REQUESTED** [review]

**Comment by mcollovati** [comment]
This is a huge one! Kudos!

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/HtmlComponentTester.java (hunk 4 above)
Comment (caalador) [reply vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/HtmlComponentTester.java:40]: Could be, container just overrides that as the HtmlContainers are HasText

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java (hunk 8 above)
Comment (caalador) [reply vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java:38]: I think so

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java (hunk 9 above)
Comment (caalador) [reply vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/OrderedListTester.java:49]: yes

File: vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/UnorderedListTester.java (hunk 11 above)
Comment (caalador) [reply vaadin-testbench-unit/src/main/java/com/vaadin/flow/component/html/testbench/UnorderedListTester.java:49]: yes

**Review by mcollovati — APPROVED** [review]

**Pull request merged**